---
title: "FigS4"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(tidyverse)
```

## Fig S4

### loading data 
```{r}
load(file = "Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/Analyses_rnaseq_recognition/results/mixed_effect_models/htseq_manual/samples.RDATA")
load(file = "Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/for_paper_counts/rMATS/lmfit_rMATS_CTGrepeat.RData")
load(file = "Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/for_paper_counts/rMATS/lmfit_rMATS_CTGrepeat_values.RData")
load(file = "Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/for_paper_counts/rMATS/rMATSSE.MATS.JC_filtered.RDATA")
```

```{r}
#cleanup exon names
lm_fit_values[["V2Mode_SE"]][,"exon_short"] <- do.call(rbind, str_split(lm_fit_values[["V2Mode_SE"]][,"exon"], "_", 2))[,2]
lm_fit_values[["V2Mode_SE"]][,"exon_short"] <- gsub("_nt", "nt", lm_fit_values[["V2Mode_SE"]][,"exon_short"])
```

### A volcano plot 
```{r}

df <- data.frame(p.value = lm_fit_values[["V2Mode_SE"]][,"p.val"], FDR = lm_fit_values[["V2Mode_SE"]][,"FDR"], effect = lm_fit_values[["V2Mode_SE"]][,"Estimate"])

ggplot(df, aes(x = effect*100, y = -log10(p.value)))+
      geom_label_repel(label = ifelse(lm_fit_values[["V2Mode_SE"]][,"exon"] %in%  lm_fit_values[["V2Mode_SE"]][,"exon"][order(lm_fit_values[["V2Mode_SE"]][,"p.val"])][1:5], lm_fit_values[["V2Mode_SE"]][,"exon_short"], ""), max.overlaps = 250, min.segment.length = 0.01, force = 50, color = "black", size = 3, nudge_y = 0.25, segment.size = 0.25) +
      geom_point(size= 1, col = ifelse(df$FDR < 0.05, "black","grey"))+
      xlab("Effect size (PSI/100 CTGs)")+
      ylab("-10log(p-value)")+
      theme(legend.position = "none", aspect.ratio = 4/3)  +
      scale_x_continuous(limits = c(-0.1, 0.1), 
                         breaks = seq(-0.1, 0.1, 0.025), 
                         labels = c(scales::comma(seq(-0.1, -0.025, 0.025), accuracy = c(0.01)), scales::comma(0, accuracy = c(1)), scales::comma(seq(0.025, 0.1, 0.025), accuracy = c(0.01))), 
                         sec.axis = dup_axis(breaks = derive(), labels = derive(), name = NULL))+
      scale_y_continuous(limits = c(0, 7), 
                         breaks = seq(0, 7, 1),
                         sec.axis = dup_axis(breaks = derive(), labels = derive(), name = NULL))+
      theme_classic()
                     
ggsave(file = "Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/for_paper_counts/Figs_for_paper/FigS4_CTG_PSI.png", width = 6, height = 8)

# The number of genes with an FDR > 0.05
sum(lm_fit_values[["V2Mode_SE"]][,"FDR"] < 0.05)
# of which with a positive correlation
sum(lm_fit_values[["V2Mode_SE"]][,"FDR"] < 0.05 & lm_fit_values[["V2Mode_SE"]][,"Estimate"] > 0 )
# of which with a negative correlation
sum(lm_fit_values[["V2Mode_SE"]][,"FDR"] < 0.05 & lm_fit_values[["V2Mode_SE"]][,"Estimate"] < 0 )

#write table with IDs to use for gene enrichment analysis

write.table(lm_fit_values[["V2Mode_SE"]],
            file = "Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/for_paper_counts/Figs_for_paper/TableS4_PSIs_CTG.csv", 
          quote = FALSE, 
          row.names= FALSE,
          col.names= FALSE
   )
```


### B example plots
```{r}

gene_plots <- list()
for (exon_ID in lm_fit_values[["V2Mode_SE"]][,"exon"][order(lm_fit_values[["V2Mode_SE"]][,"p.val"])][1:5]){

outcome <- "V2Mode"
      # for V2Mode, create a scatter plot with expression level vs CTG-repeat length
 df2 <- data.frame(as.numeric(samples[,outcome][!is.na(samples[,outcome])]), SE.MATS.JC_filtered[["IncLevel1"]][exon_ID,], samples[,"Visit"][!is.na(samples[,outcome])], samples[,"PatientID"][!is.na(samples[,outcome])])
names(df2) <- c(outcome,"counts","Visit","PatientID")
df2 <- df2[order(df2$Visit),]
gene_plots[[exon_ID]] <- ggplot(df2, aes_string(x=outcome, y="counts",group="PatientID")) +
        ggtitle(paste(lm_fit_values[["V2Mode_SE"]][lm_fit_values[["V2Mode_SE"]][,"exon"] == exon_ID,"exon_short"])) +
        xlab("CTG repeat length") +
        ylab("PSI") +
        geom_point(size = 1.5, col=ifelse(df2$Visit == "V2", "blue","red")) +
        scale_x_continuous(breaks=seq(0,1000,200), 
                           limits = c(0,1000),
                           sec.axis = dup_axis(breaks = derive(), labels = NULL, name = NULL))+
        scale_y_continuous(breaks=seq(0, 1, 0.2), 
                           limits = c(0,1),
                           sec.axis = dup_axis(breaks = derive(), labels = NULL, name = NULL))+
  theme_classic() +
  theme(panel.grid.major.x = element_line(size = 0.25, color = "grey"))+
  theme(panel.grid.major.y = element_line(size = 0.25, color = "grey"))
}

plots <- plot_grid(plotlist = gene_plots, ncol = 1 )
ggsave(plots, file ="Z:/Analysis_projects/RvC_recognition_rnaseq_diff_expr/for_paper_counts/Figs_for_paper/FigS4_CTG_PSIplots.png", height = 20, width = 8, dpi = 600 )

```